mautic/mautic
# Speed up segment detail page

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | features
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | /
| Related developer documentation PR URL | /
| Issue(s) addressed                     | /




#### Description:
The main goal of this PR is to remove unnecessary subquery that is causing timeouts on segment detail pages. Here is the full query with the subquery:
```sql
SELECT DATE_FORMAT(t.date_added, '%Y-%m-%d') AS date,
       (COUNT(DISTINCT t.lead_id)) as count
FROM mautic_lead_event_log t
WHERE (t.object = 'segment')
  AND (t.bundle = 'lead')
  AND (t.action = 'added')
  AND (t.object_id = '446')
  AND (t.date_added BETWEEN '2020-06-21 00:00:00' AND '2020-07-21 23:59:59')
  AND (NOT EXISTS
         (SELECT null
          FROM mautic_lead_event_log el
          WHERE (el.lead_id = t.lead_id)
            AND (el.object = 'segment')
            AND (el.bundle = 'lead')
            AND (el.action = 'removed')
            AND (el.object_id = 446)
            AND (DATE_FORMAT(el.date_added, '%Y-%m-%d') = DATE_FORMAT(t.date_added, '%Y-%m-%d'))
            AND (el.date_added > t.date_added)))
GROUP BY DATE_FORMAT(t.date_added, '%Y-%m-%d')
ORDER BY DATE_FORMAT(t.date_added, '%Y-%m-%d') ASC
LIMIT 31; 
```

Here is a discussion about why the subquery should be removed:
https://github.com/mautic/mautic/pull/7285/files#r458154398

However, when writing the functional test for this I notice an issue in the chart data. The contacts added were visible 1 day prior to the actual date added. So I fixed that as well.

<img width="1154" alt="Screen Shot 2020-09-17 at 16 40 18" src="https://user-images.githubusercontent.com/1235442/93874723-17283a00-fcd4-11ea-898a-fd99a35fbfbb.png">

Another issue I noticed was that the counts were done twice by calling the init method twice. I removed the unnecessary call. If you compare the Symfony dev bar shows we saved 4 SQL requests with this. It used to make 53 queries and it's doing 49 now. See https://github.com/mautic-inc/mautic-cloud/pull/1013/commits/d326e6c1fdeafd4895af5079941ce045d28c9f2f

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. It'd be great if you have some date in your testing instance already so you could see them already in the chart
2. If you don't have any segments with contacts then create some and add contacts to them. Doesn't matter whether with a filter or manually.
3. Notice the issue on the screenshot above.

#### Steps to test this PR:
0. Load up [this PR](https://mautibox.com)
1. Refresh the page. The added and removed contact counts should be aligned with the total counts. The site should load faster especially for big datasets.
2. Test also different date ranges. For example if you select the same from and to date then the time unit will change from days to hours. Similarly, if you select a span of 2 months then the time unit should change to weeks. All should show the correct data.
3. Test also with different timezones.

#### Other areas of Mautic that may be affected by the change:
1. All charts in the UI ⚠️ (Dashboard charts, Email detail page, Landing page, Campaign detail page, ...)

#### List deprecations along with the new alternative:
1. None

#### List backwards compatibility breaks:
1. None
